### ISSUE TITLE
clang++ behaves non-standard with crtp template 

### ISSUE THREAD

**Issue by pmqtt** [llvm/llvm-project#55855]
Description:
The following code should be compile. This is valid C++ code and the g++ compile it.

```cpp
template<class T>
struct Base{
    using PARAMETER = typename T::PARAMETER;
};

template<class T,class Y>
struct Derivated : public Base<Derivated<T,Y>>{
    using PARAMETER = Y;
};

int main(int argc,char **argv){
    Derivated<int,float> k;

}
```

Produces following output:
```console
error: no type named 'PARAMETER' in 'Derivated<int, float>'
```

Version of clang++ is:
```
clang version 13.0.1
```

**Comment by shafik** [comment]
I do not believe this code is valid see this [Stackoverflow question](https://stackoverflow.com/q/6006614/1708801).

Which gcc version does this compile with, I can not find a version on [godbolt that does not produce a diagnostic for this code](https://godbolt.org/z/1xssjzfbc)

**Comment by pmqtt** [comment]
Sry,I've oversimplified the problem!

Now here is the code sample that compiles under g++ version 12.1.0

```
template<class T>
struct Base{
    template<class Y>
    requires(std::is_same<typename T::PARAMETER,Y>::value) void foo(const Y & x){ }
};

template<class T,class Y>
struct Derivated : public Base<Derivated<T,Y>>{
    using PARAMETER = Y;
};
```

**Comment by pmqtt** [comment]
Oh I'm so sorry... I took a look on godbolt and see the same error... I think gcc on manjaro is compiled with Flags that allowed my code.

**Comment by shafik** [comment]
The [godbolt I am using shows success for gcc but failure for clang](https://godbolt.org/z/bq9nGxceM).

Let me know if I am missing something.

**Comment by pmqtt** [comment]
It's to late in germany... I forgot the c++20 flag. 

My question, is it a bug in clang or is this code non-standard?

**Comment by shafik** [comment]
I believe this is https://github.com/llvm/llvm-project/issues/44178 

CC @erichkeane

**Comment by erichkeane** [comment]
> I believe this is #44178
> 
> CC @erichkeane

Yep, I'm about 95% sure this is a duplicate of that one, requires clauses aren't properly deferred in evaluation in Clang at the moment, but I have a review opened just yesterday that fixes this one.  There are some pending problems with 'constrained friends' that aren't fixed by mine, but the situation is improved.

**Comment by royjacobson** [comment]
Works after https://reviews.llvm.org/D126907, marking as duplicate and closing.
